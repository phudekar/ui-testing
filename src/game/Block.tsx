import React from 'react';
import Position from '../lib/position';
import Block from '../lib/block';

const BlockComponent = ({ block, onReveal, toggleFlagged, disabled = false }: BlockProps) => {
    const position: Position = block?.position || {};
    return (
        <div className={`block${block?.revealed ? ' revealed' : ''}`}
            data-testid={`block-${position.row}-${position.column}`}
            onClick={() => !disabled && onReveal && onReveal()}
            onContextMenu={e => {
                e.preventDefault();
                !disabled && toggleFlagged && toggleFlagged()
            }}>
            {
                block.revealed ? (
                    block?.exploded ? <div className="bomb" />
                        : <div className="nearby-bombs">{block.nearbyBombs || ''}</div>
                ) : (block?.flagged ? <div className="flag" /> : <div className="mine" />)
            }

        </div>
    )
}

export interface BlockProps {
    block: Block,
    onReveal?: Function,
    toggleFlagged?: Function
    disabled?: boolean;
}

export default BlockComponent;