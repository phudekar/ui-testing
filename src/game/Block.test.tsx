import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BlockComponent from './Block';
import Block from '../lib/block';
import Position from '../lib/position';

describe('BlockComponent', () => {
    const position = new Position(2, 1)

    it('should not be revealed by default', () => {
        const block = new Block(position);
        const { getByTestId } = render(<BlockComponent block={block} />);

        const root = getByTestId(`block-${position.row}-${position.column}`);

        expect(root.classList).toContain('block');
        expect(root.classList).not.toContain('revealed');
    })

    it('should call onRevealed after left click', () => {
        const block = new Block(position);
        const reveal = jest.fn();

        const { getByTestId } = render(<BlockComponent block={block} onReveal={reveal} />);

        const root = getByTestId(`block-${position.row}-${position.column}`);
        fireEvent.click(root);

        expect(reveal).toHaveBeenCalled();
    })

    it('should call toggleFlagged after right click', () => {
        const block = new Block(position);
        const toggleFlagged = jest.fn();

        const { getByTestId } = render(<BlockComponent block={block} toggleFlagged={toggleFlagged} />);

        const root = getByTestId(`block-${position.row}-${position.column}`);

        fireEvent.contextMenu(root);

        expect(toggleFlagged).toHaveBeenCalled();
    })

    it('should add revealed class by after left click', () => {
        const block = new Block(position);
        block.reveal([]);

        const { getByTestId } = render(<BlockComponent block={block} />);

        const root = getByTestId(`block-${position.row}-${position.column}`);

        expect(root.classList).toContain('revealed');
    })

    it('should show flag after right click', () => {
        const block = new Block(position);
        block.flag();

        const { container } = render(<BlockComponent block={block} />);

        expect(container.querySelector('.flag')).toBeVisible();
    })

    it('should show bomb if revealed and exploded', () => {
        const { container } = render(
            <BlockComponent block={{ revealed: true, exploded: true }} />
        );

        expect(container.querySelector('.bomb')).toBeVisible();
    })

    it('should show number of nearByBombs if revealed', () => {

        const { queryByText } = render(
            <BlockComponent block={{ revealed: true, nearbyBombs: 2 }} />
        );

        expect(queryByText("2")).toBeVisible();
    })
})
