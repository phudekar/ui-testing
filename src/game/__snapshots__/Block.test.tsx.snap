// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockComponent should render Block with default state 1`] = `
<div
  class="block"
  data-testid="block-2-1"
>
  <div
    class="mine"
  />
</div>
`;

exports[`BlockComponent should show bomb if revealed and exploded 1`] = `
<div
  class="block revealed"
  data-testid="block-undefined-undefined"
>
  <div
    class="bomb"
  />
</div>
`;

exports[`BlockComponent should show flag after right click 1`] = `
<div
  class="block"
  data-testid="block-2-1"
>
  <div
    class="flag"
  />
</div>
`;

exports[`BlockComponent should show number of nearByBombs if revealed 1`] = `
<div
  class="block revealed"
  data-testid="block-undefined-undefined"
>
  <div
    class="nearby-bombs"
  >
    2
  </div>
</div>
`;
